dot/telemetry/telemetry_test.go

```diff
@@ -113,6 +113,32 @@ func TestHandler_SendMulti(t *testing.T) {
 	require.Contains(t, string(actual[3]), string(expected[3]))
 }
 
+func TestListenerConcurrency(t *testing.T) {
```

Comment:
Thanks for the test. This helped me to debug the issue.

You can take a lock at the connection level which is optimized than taking the lock at the handler level because you can send multiple messages concurrently to different connections.
This issue is `connection` in `Handler.connections` is value and not pointer. Hence it creates multiple instances of lock for the same connection.

Add lock at the connection level but change `Handler.connections` to a pointer.
```
type telemetryConnection struct {
	wsconn    *websocket.Conn
	verbosity int
	sync.Mutex
}

// Handler struct for holding telemetry related things
type Handler struct {
	msg         chan Message
	connections []*telemetryConnection
}

func (t *Handler) startListening() {
	for {
		msg := <-t.msg
		go func() {
			for _, v := range t.connections {
				v.Lock()
				_ = v.wsconn.WriteMessage(websocket.TextMessage, msgToBytes(msg))
				v.Unlock()
			}
		}()
	}
}

```

Also, I think users should get notified in logs if they are unable to write messages to the telemetry server. So that they can take action accordingly.

Reply:
Thanks for debugging this, it makes sense now.  I've updated to fix it, and added logging to catch any error and warn the user.